Bootstrap v5

smithdc1 opened the pull request:
Fixes #7327 

This is a WIP patch to upgrade the templates to Bootstrap v5. I've tried to chunk the commits up to help review, this will change as this will quite a number of files and lines of code. 

I feel like I've made good progress so far. Only major issue I've found so far is tooltips on the dropdown button. The solution on bootstrap 4 (which I've used here) no longer works on bootstrap 5. Another alpha release is due of bootstrap 5 shortly, if it is not fixed in that I can try approaching the issue tracker. 

To do (there is much more than this, but here are things I've changed but need to go back and look at):
- [ ] Fix tooltip on split dropdown button
- [x] Change the js file to the bundled version (can then drop link to popper.js)

smithdc1 committed:
Updated static files to BS-v5 & added popper.js

smithdc1 committed:
Navbar and top buttons
* navbar-expand-md to stop navbar growing when dropdown opens.
* Split dropdown doesn't work with tooltip - looks like a bs5 bug (same layout works with bs4)
* Need to check button colours
* Need to look at colour of dropdown links

smithdc1 committed:
Page header
* BS3 had lots of top (20px) and bottom (10px) margin to push page down. Now added top margin on the div.
* Font has style has changed with bs5. But is slightly custom as font weight is over riden. Personally think it looks ok.

smithdc1 committed:
Pagination (numbers.html)

smithdc1 committed:
HTML/Raw data forms

* form-horizontal is no longer used. Need to use the grid.

aligned tabs right

Raw Data Form
* Use grid / rows to space form. form-group has been deprecated in favour of utility classes.

Replaced `well` class (background colour/border) with bootstrap utility classes

smithdc1 committed:
Changed form-group to mb-3
* form-group is deprecated in bootstrap 5 and is replaced by utility classes.

smithdc1 committed:
Horizontal forms

* checkbox are now fully custom by default in Bootstrap5 - removed previous tweaks.
* control-label --> col-form-label
* errors now need `invalid-feedback` on the text and `is-invalid` on the input
* new `form-text` class for help text
* New html layout and css classes for radio/multiple checkbox. Added work around for errors and inline (see comments in code).

smithdc1 committed:
Fixed tests following horizontal form changes
* error class has changed to `invalid-feedback`

smithdc1 committed:
Vertical forms

smithdc1 committed:
Filters

* Used bootstrap v5 icon pack

smithdc1 committed:
Inline Forms

smithdc1 committed:
Admin
* Updated glyphicons to Bootstrap 5 icons

carltongibson commented:
Hi @smithdc1. Thanks for this! Super start.

smithdc1 commented:
Thanks 😄 

It's coming along well, currently wrangling with the collapse menu in the docs. 

Got my eye on the upcoming releases, seems like they made good progress on alpha 2 in the last few days. I can already see some breaking changes in the next release but they should be incremental changes to the current changes. (e.g. rename of `sr-only`)

smithdc1 committed:
Docs

* Fixed sidebar collapse (links now work)
* Changed fontawsome icons to bootstrap icones

smithdc1 committed:
Remove duplicate <p> tags
* render_markdown returns its content in a <p> tag, so don't need to wrap again.

smithdc1 committed:
remove fontawesome

smithdc1 committed:
Used bootstrap5 bundled js
* No need to add popper.js as a separate source

smithdc1 commented:
So I'm fairly pleased with where I've got this to. It's a bit of a waiting game now to see the new releases come out and see if there are any more breaking changes, and to see if the drop-down tool-tip gets fixed. Can also review the remaining css in `bootstrap-tweaks` and `default` to see if any further rationalisation can take place here. 

One question I do have is for the filter form. The filter form looks a bit odd, it looks like a template from `django-filter`, what's going on here? (I think I know where this is headed... 😄 )

carltongibson commented:
> One question I do have is for the filter form. The filter form looks a bit odd, it looks like a template from django-filter, what's going on here? (I think I know where this is headed... 😄 )

Chuckle. 😀

Erm... the form is generated by django-filter but is it rendered with crispy forms? (A while since I looked in here...) If you throw up a few links to exact places I'm sure I can get the relevant grey-cells lined up again. 😝

smithdc1 commented:
Ok - let's leave it some time and I'll dig deeper at some point in the coming months. There is going to be plenty of time before Bootstrap 5 reaches a stable release. 

There may be a `crispy-bootstrap-5` template pack by then (template work is done, I need to find the effort to put it into a package, write docs, setup c/i etc)

smithdc1 commented:
Note (to self really).

#30170 on bootstrap 5 changes feedback messages. Html layout and CSS classes will need to be changed.

Currently this ticket is on the inbox for v5 alpha 2.

joshjung commented:
Looking forward to this!

smithdc1 commented:
> Looking forward to this!

Thanks -- It's coming but bootstrap 5 development itself is steadily paced. For your need I'm not sure I have a good answer, maybe you could have your own fork with this work, noting it is un-reviewed (but if you're in the market to have a look 😉 ) I'm not sure an alpha version is better than an old one from a security perspective 🤷 .

smithdc1 commented:
[Bootstrap5 beta 1](https://blog.getbootstrap.com/2020/12/07/bootstrap-5-beta-1/) is now out. I'll be looking to test my memory by revisiting this _soon_ :tm:.

smithdc1 committed:
Bump Bootstrap to v5.0.0-beta3

smithdc1 committed:
Namespaced data attributes

Changed in beta 1 see - https://getbootstrap.com/docs/5.0/migration/#javascript-2

smithdc1 committed:
Changed hide to d-none

The hide class in bootstrap 2 added display:none.

This is done via d-none in v5 see.

https://getbootstrap.com/docs/5.0/utilities/display/#notation

smithdc1 committed:
Changed left/right to start/end

Changed in beta 1. See https://getbootstrap.com/docs/5.0/migration/#utilities-2

smithdc1 committed:
Fixed dropdown button

smithdc1 committed:
Fixed test test_rendering_nested_fields_with_not_mappable_value

smithdc1 commented:
Hi all, 

So I think this is ready for review. I've left the commit history to show how I achieved this in smaller steps, but I'm not sure if they could be better presented to make the review easier. I am not used to writing such big patches, I also apologise for the size, but I'm not sure there is much I can do. Any guidance would be appreciated. 

I'm not sure anyone here is going to have the capacity to review this any time soon (is any one else finding 2021 harder than 2020?). I was wondering if it could be worth trying to widen the net of potential reviewers of this with a post on Twitter?  I fully expect this to be far from perfect and even if it's a partial review here and there it would get us closer to the finish line. (I'm very unsure, but I am trying to think of ways of making progress in a sustainable way).

jscaux commented:
@smithdc1 big thumbs up on this, I'm very keen on using BS5 in DRF. Hopefully this will get reviewed and merged soon. You're doing just great with 2021.

carltongibson commented:
@jscaux if you were to install this locally and look for issues, that would be a helpful step. 👍

jscaux commented:
@carltongibson will do; currently busy with moving a separate big stack to BS5. Since DRF is part of the backend, it's a road I have to go down anyway.

jscaux commented:
@carltongibson I've managed to install this locally and run a very basic setup without issues. The bootstrap version could be bumped up to 5.0.1.

carltongibson commented:
Ok, thanks for the feedback @jscaux. Let us know if you hit other issues.

lucasdavid commented:
Has this been abandoned? DRF could really improve with bs5 :(

tomchristie commented on rest_framework/templates/rest_framework/base.html:
```diff
@@ -79,38 +77,37 @@
+                  <div class="btn-group">
+                    <button class="btn btn-primary" data-bs-toggle="tooltip" href="{{ request.get_full_path }}" rel="nofollow" title="Make a GET request on the {{ name }} resource">GET</button>
+                    <button class="btn btn-primary dropdown-toggle dropdown-toggle-split" data-bs-toggle="dropdown">
+                      <span class="sr-only" data-bs-toggle="tooltip" title="Specify a format for the GET request">Toggle Dropdown</span>
```
This renders "Toggle Dropdown" next to the dropdown control. I'm just going to remove this line for now.
If someone wants to add an `sr-only` here that doesn't render visibly, then let's tackle that separately.

```suggestion
```

lovelydinosaur committed:
Update rest_framework/templates/rest_framework/base.html

tomchristie commented:
Okay, so I've been reviewing this. I've made one fix to the templates. Also I think it's worth comparing the two versions here...

## Existing Browsable API

<img width="758" alt="Screenshot 2021-12-10 at 16 07 38" src="https://user-images.githubusercontent.com/647359/145606048-ed76f3b3-f4cb-4be3-ad66-56318d2b080f.png">

<img width="753" alt="Screenshot 2021-12-10 at 16 07 48" src="https://user-images.githubusercontent.com/647359/145606084-6cfc0a72-2a11-4a2f-b281-702cc2bdf7ba.png">

<img width="753" alt="Screenshot 2021-12-10 at 16 07 55" src="https://user-images.githubusercontent.com/647359/145606105-cd7fe474-aa41-4f24-bfab-733538ddf16d.png">

## Updated Browsable API

<img width="817" alt="Screenshot 2021-12-10 at 16 09 53" src="https://user-images.githubusercontent.com/647359/145606177-fbb843a0-c4a8-4e31-a029-61839382d211.png">

<img width="788" alt="Screenshot 2021-12-10 at 16 09 47" src="https://user-images.githubusercontent.com/647359/145606189-b726bcfd-5704-434e-ba43-0733e8a62942.png">

<img width="773" alt="Screenshot 2021-12-10 at 16 09 40" src="https://user-images.githubusercontent.com/647359/145606209-7abdeff0-793f-45ec-ba98-9d326d7f2bb8.png">

---

## Existing Admin Renderer

<img width="753" alt="Screenshot 2021-12-10 at 16 06 52" src="https://user-images.githubusercontent.com/647359/145606406-d77d9e66-3fa5-4473-9627-de45cc58f21f.png">

<img width="755" alt="Screenshot 2021-12-10 at 16 06 57" src="https://user-images.githubusercontent.com/647359/145606438-41f8f31b-6c6b-4231-bab7-99644eba4f55.png">

<img width="758" alt="Screenshot 2021-12-10 at 16 07 10" src="https://user-images.githubusercontent.com/647359/145606470-243929db-b208-4560-928d-429ee6447ed8.png">

## Updated Admin Renderer

<img width="776" alt="Screenshot 2021-12-10 at 16 10 09" src="https://user-images.githubusercontent.com/647359/145606519-d1a46469-30d7-4a1f-98df-8a1809145a65.png">

<img width="773" alt="Screenshot 2021-12-10 at 16 10 14" src="https://user-images.githubusercontent.com/647359/145606538-098537ba-1216-40b2-bd48-230e664fe6ce.png">

<img width="804" alt="Screenshot 2021-12-10 at 16 10 22" src="https://user-images.githubusercontent.com/647359/145606559-bc4d51c3-0858-4e84-ad4a-a824b8561db2.png">

Personally I'm finding the styling changes a bit jarring here, but perhaps that's because I'm used to the existing versions?

It'd be nice to update, and this is a *fantastic* bit of work getting us there, but it's a shame that the style looks a bit off to me after the update.

Any thoughts, folks?

auvipy commented:
actually newer version of twbs follow newer design trend AFAIK.

lucasdavid commented:
Maybe it looks a bit off because the new buttons are a little lighter and brighter, while DRF's components (e.g. navbar) are darker and opaque. I always though there was quite a change in tone from bootstrap v3 to v5.

tomchristie commented:
So I think the obvious things that need fixing are...

* The button sizes after this change are significantly bigger, and they don't fit the layout so well.
* The changed `+` and `x` icons for the create/delete buttons are just a bit blergh.
* The brighter palette perhaps doesn't suit the rest of the interface style quite as well, but I don't think that's a massive problem.

lucasdavid commented:
Breadcrumbs are quite different too. Was that intentional? I think `.bg-light .rounded` classes could be added here.

JJZolper commented:
I left a comment: https://github.com/encode/django-rest-framework/pull/7411/commits/5577d3632e1b20466dab1c4bbcc2d7e7bfb15294#r805194699

I think anywhere there is .sr-only that has to be removed and the v5 class used.

auvipy commented:
as this is staled, should we try a smaller incremental version with bootstrap v4 instead?

baseplate-admin commented:
> as this is staled, should we try a smaller incremental version with bootstrap v4 instead?

or maybe we should have our own CSS like `django-admin` 👀

auvipy commented:
we can revamp it with newer https://blog.getbootstrap.com/2023/05/30/bootstrap-5-3-0/ which has new colors and many improvements.

baseplate-admin commented:
Hi, can this be rebased?

luanfonceca commented:
Hi, @smithdc1

I want to help you finish this PR. I have extensive experience with Django templates and Bootstrap 5 (I just did a similar migration on a client from v3 to v5.3).

I plan to pull your branch, update this to Bootstrap v5.3, and keep going page by page, ensuring everything works seamlessly.
In the best scenario, we would do this incrementally, keeping the existing Bootstrap v3 stylesheets, js, and jQuery in use while we're updating each piece, but as this is an open-source library, I don't think we could do this.

Are you available for a quick async talk on the IRC channel, or do you prefer to discuss the next steps here?

smithdc1 commented:
Hi @luanfonceca  -- I'm afraid I don't have time to work on this patch right now. It's been ~3 years since I've looked at this so I'd be starting from scratch to pick this up again. I think it would be amazing if you have capacity to update and address the review comments!

luanfonceca commented:
Yeah, there's no need to worry about it.
I had that in mind already.

Thanks

luanfonceca commented:
Hi @auvipy, I can take over this migration, I saw some conversation about migrating to v4 before trying the v5, do you know how that's going?

Last week did some attempts and can share where I'm at with you.

luanfonceca commented:
Here's a short video showing how far I got with this last week; I had a couple of issues, and the layout still needs some improvements.

I can open a new issue up for discussion, along with my ideas about this, as there were several attempts related to this migration. I think we need time to decide what we want to do and how, as well as the impact of this migration on third-party apps that use the DRF browseable API page.

https://github.com/encode/django-rest-framework/assets/1490875/d2357af3-bfab-44fa-8b2a-f12069a8d99a

martinberoiz commented:
Could this maybe be released as some alpha version for some future release? Or is it too risky? I'm interested in BS5 too

rmiyazaki6499 commented:
Any update on this?

We recently had a pen test done and they told us that we need to upgrade Bootstrap due to security vulnerabilities

<img width="862" alt="Screenshot 2567-07-30 at 15 05 55" src="https://github.com/user-attachments/assets/40f40c2e-0dfb-43e1-9d4e-fbfd4b08c43c">

It seems this has been on the back burner for a while, any plans on updating? @luanfonceca

luanfonceca commented:
Hey @rmiyazaki6499, sorry about that.

I tried to migrate that and got it in a really good state, but it is not production-ready.

The major problem I faced was that both the Docs and the user-facing (the front-end bundle that's pip installed) share the same front-end codebase, so to make the whole migration (DRF browseable API and docs), it needs to be done together.

I can get back to this if we are open to starting with just the user-facing first and migrating it separately. I'd ask that for two reasons:
1. It's probably where the vulnerabilities are more critical to get fixed
2. It's less work that could be better reviewed and launched.

samanthagoodhue commented:
Thanks for trying to get this fixed @luanfonceca. In the same situation as @rmiyazaki6499 with pentest findings. Again relatively low risk, but would be great if you were able to fix 🙏

anthonyrom commented:
Hi @luanfonceca thanks for your work on this. I'd be happy to try and pick up where you've left off if you don't have the time to work on this in the near term.

auvipy commented:
you can take this and fix the merge conflicts first and make it working again

luanfonceca commented:
> Hi @luanfonceca thanks for your work on this. I'd be happy to try and pick up where you've left off if you don't have the time to work on this in the near term.

Go for it!

It was really hard to upgrade to Bootstrap v5, mainly because the same Base templates used in the User-facing API are also used in the Docs. Instead of upgrading just one template and a few components, I got caught up in the whole Docs website codebase, which was time-consuming and brought up several bugs and broken pieces of UI.

Creating a new base template for the user-facing API and upgrading only the HTML used there will make this migration easier.

anthonyrom commented:
It looks like [Bootstrap v5 upgrades will not be accepted for the Docs](https://github.com/encode/django-rest-framework/pull/9404#issuecomment-2346694239). Is this dead in the water until the "unified doc style" gets implemented (i.e. and the Docs site + user-facing API no longer use the same base templates)? Probably a question for @tomchristie

auvipy commented:
@tomchristie what do you think if I start working on this?

browniebroke commented:
Closing this as it's too out of date.
